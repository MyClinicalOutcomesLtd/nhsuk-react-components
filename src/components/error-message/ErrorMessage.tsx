import classNames from 'classnames';
import React, { HTMLProps, useContext, useEffect } from 'react';
import FormGroupContext from '../form-group/FormGroupContext';

interface ErrorMessageProps extends HTMLProps<HTMLSpanElement> {
  visuallyHiddenText?: false | string;
}

const ErrorMessage: React.FC<ErrorMessageProps> = ({
  id,
  className,
  visuallyHiddenText,
  children,
  ...rest
}) => {
  const { inputID, setError } = useContext(FormGroupContext);
  const elementID = id || (inputID ? `${inputID}--error-message` : undefined);

  useEffect(() => {
    if (children) setError(true);
    // return setError(false);
  }, [children]);

  return (
    <span className={classNames('nhsuk-error-message', className)} id={elementID} {...rest}>
      {visuallyHiddenText !== false ? (
        <span className="nhsuk-u-visually-hidden">{visuallyHiddenText}</span>
      ) : null}
      {children}
    </span>
  );
};

ErrorMessage.defaultProps = {
  visuallyHiddenText: 'Error: ',
  role: 'alert',
};

export default ErrorMessage;
