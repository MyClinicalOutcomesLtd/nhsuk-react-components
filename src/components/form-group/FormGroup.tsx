import classNames from 'classnames';
import React, { HTMLProps, useState } from 'react';
import FormGroupContext from './FormGroupContext';

export interface FormGroupProps extends HTMLProps<HTMLDivElement> {
  error?: boolean;
  disableErrorLine?: boolean;
  disableErrorFromComponents?: boolean;
  _exposeContext?: boolean;
}

const FormGroup: React.FC<FormGroupProps> = ({
  error: errorProp,
  children,
  className,
  disableErrorLine,
  disableErrorFromComponents,
  _exposeContext,
  ...rest
}) => {
  const [error, setError] = useState<boolean>(errorProp);
  const [inputID, setInputID] = useState<string>();

  const hasError =
    typeof error === 'boolean' ? error : !disableErrorFromComponents && !disableErrorLine && error;

  return (
    <div
      className={classNames('nhsuk-form-group', { 'nhsuk-form-group--error': hasError }, className)}
      {...rest}
    >
      {_exposeContext ? (
        <FormGroupContext.Provider
          value={{
            isInFormGroup: true,
            inputID,
            error,
            setError,
            setInputID,
          }}
        >
          {children}
        </FormGroupContext.Provider>
      ) : (
        children
      )}
    </div>
  );
};

FormGroup.defaultProps = {
  _exposeContext: true,
};

export default FormGroup;
