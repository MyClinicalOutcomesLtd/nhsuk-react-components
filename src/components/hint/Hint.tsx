import classNames from 'classnames';
import React, { HTMLProps, useContext } from 'react';
import FormGroupContext from '../form-group/FormGroupContext';

type HintProps = HTMLProps<HTMLDivElement>;

const Hint: React.FC<HintProps> = ({ className, id, ...rest }) => {
  const { inputID } = useContext(FormGroupContext);
  const elementID = id || (inputID ? `${inputID}--hint` : undefined);
  return <div className={classNames('nhsuk-hint', className)} id={elementID} {...rest} />;
};

export default Hint;
