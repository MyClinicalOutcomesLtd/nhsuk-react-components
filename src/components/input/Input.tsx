import classNames from 'classnames';
import React from 'react';
import useFormGroup from '../../util/hooks/UseFormGroup';
import { FormElementProps, FormGroupConsumer } from '../../util/types/FormTypes';
import { InputWidth } from '../../util/types/NHSUKTypes';

interface InputProps extends FormElementProps<HTMLInputElement> {
  width?: InputWidth;
  disableErrorLine?: boolean;
}

const Input: React.FC<InputProps> = React.forwardRef((props, ref) => {
  const group = useFormGroup<InputProps>(FormGroupConsumer.INPUT, props);
  const { width, className, error, ...rest } = group.renderProps;

  return (
    <group.Wrapper {...group.wrapperProps}>
      {group.LabelBlock}
      <input
        className={classNames(
          'nhsuk-input',
          { [`nhsuk-input--width-${width}`]: width },
          { 'nhsuk-input--error': error },
          className,
        )}
        ref={ref}
        {...rest}
      />
    </group.Wrapper>
  );
});

Input.displayName = 'Input';

Input.defaultProps = {
  type: 'text',
};

export default Input;
