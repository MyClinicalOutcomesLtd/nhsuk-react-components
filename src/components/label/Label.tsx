import classNames from 'classnames';
import React, { HTMLProps, useContext } from 'react';
import { NHSUKSize } from '../../util/types/NHSUKTypes';
import FormGroupContext from '../form-group/FormGroupContext';

interface LabelProps extends Omit<HTMLProps<HTMLLabelElement>, 'size'> {
  bold?: boolean;
  isPageHeading?: boolean;
  size?: NHSUKSize;
}

const BaseLabel: React.FC<LabelProps> = ({ id, className, bold, size, isPageHeading, ...rest }) => {
  const { inputID, getLabelID } = useContext(FormGroupContext);
  return (
    // eslint-disable-next-line jsx-a11y/label-has-associated-control
    <label
      id={getLabelID(id)}
      htmlFor={inputID}
      className={classNames(
        'nhsuk-label',
        { 'nhsuk-label--s': bold && !size },
        { 'nhsuk-label--xl': isPageHeading && !size },
        { [`nhsuk-label--${size}`]: size },
        className,
      )}
      {...rest}
    />
  );
};

const Label: React.FC<LabelProps> = ({ isPageHeading, ...rest }) => {
  if (isPageHeading) {
    return (
      <h1 className="nhsuk-label-wrapper">
        <BaseLabel isPageHeading {...rest} />
      </h1>
    );
  }
  return <BaseLabel {...rest} />;
};

export default Label;
