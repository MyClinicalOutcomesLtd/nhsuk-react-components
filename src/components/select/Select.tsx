import classNames from 'classnames';
import React from 'react';
import ConditionalFormGroup from '../../util/ConditionalFormGroup';
import useFormComponent, { FormElementProps, InputType } from '../../util/hooks/UseFormComponent';
import LabelBlock from '../../util/LabelBlock';

const Select = React.forwardRef<HTMLSelectElement, FormElementProps<HTMLSelectElement>>(
  (props, ref) => {
    const { renderProps, labelBlockProps } = useFormComponent(InputType.SELECT, props);
    const { children, className, ...rest } = renderProps;

    return (
      <ConditionalFormGroup error={props.error}>
        <LabelBlock {...labelBlockProps} />
        <select
          className={classNames('nhsuk-select', { 'nhsuk-select--error': props.error }, className)}
          {...rest}
          ref={ref}
        >
          {children}
        </select>
      </ConditionalFormGroup>
    );
  },
);
Select.displayName = 'Select';

export default Select;
