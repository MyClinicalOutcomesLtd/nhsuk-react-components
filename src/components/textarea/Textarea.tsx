import classNames from 'classnames';
import React from 'react';
import useFormGroup from '../../util/hooks/UseFormGroup';
import LabelBlock from '../../util/LabelBlock';
import { FormElementProps, InputType } from '../../util/types/FormTypes';

type TextareaProps = FormElementProps<HTMLTextAreaElement>;

const Textarea: React.FC<TextareaProps> = (props) => {
  const { FormGroup, renderProps } = useFormGroup(InputType.TEXTAREA, props);

  const { className, error, ...rest } = renderProps;

  return (
    <FormGroup error={error}>
      <LabelBlock elementId={renderProps.id} error={error} {...renderProps} />
      <textarea
        className={classNames('nhsuk-textarea', { 'nhsuk-textarea--error': error }, className)}
        {...rest}
      />
    </FormGroup>
  );
};

export default Textarea;
