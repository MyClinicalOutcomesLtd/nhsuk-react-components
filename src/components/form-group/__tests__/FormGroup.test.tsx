import React from 'react';
import { shallow } from 'enzyme';
import FormGroup from '..';
import Label from '../../label';
import ErrorMessage from '../../error-message';
import Hint from '../../hint';
import Input from '../../input';
import Textarea from '../../textarea';
import Select from '../../select';

describe('Form Group', () => {
  it('matches snapshot Form group and label', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot Form group, label and Hint', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
        <Hint>Hint: Use this box to specify more</Hint>
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot Form group label and Textarea', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
        <Hint>Hint: Use this box to specify more</Hint>
        <Textarea
          id="no-ni-reason2"
          name="no-ni-reason2"
          rows={5}
          label="Why can&#39;t you provide a National Insurance number?"
        />
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot Form group label , Textarea and Error message', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
        <Hint>Hint: Use this box to specify more</Hint>
        <ErrorMessage>At least 20 characters...</ErrorMessage>
        <Textarea
          id="no-ni-reason2"
          name="no-ni-reason2"
          rows={5}
          label="Why can&#39;t you provide a National Insurance number?"
        />
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot Form group label , Input and Error message', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
        <Hint>Hint: Use this box to specify more</Hint>
        <ErrorMessage>At least 20 characters...</ErrorMessage>
        <Input id="input-id4" width="10" />
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot Form group label , Select and Error message', () => {
    const element = shallow(
      <FormGroup>
        <Label>Please specify (others)</Label>
        <Hint>Hint: Use this box to specify more</Hint>
        <ErrorMessage>At least 20 characters...</ErrorMessage>
        <Select>
          <Select.Option>Select</Select.Option>
          <Select.Option value="1">NHS.UK frontend option 1</Select.Option>
          <Select.Option value="2">NHS.UK frontend option 2</Select.Option>
          <Select.Option value="3">NHS.UK frontend option 3</Select.Option>
        </Select>
      </FormGroup>,
    );
    expect(element).toMatchSnapshot();
  });
});
