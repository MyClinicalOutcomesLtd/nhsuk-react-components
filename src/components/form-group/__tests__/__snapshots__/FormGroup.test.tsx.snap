// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Group matches snapshot Form group and label 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
  </ContextProvider>
</div>
`;

exports[`Form Group matches snapshot Form group label , Input and Error message 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
    <Hint>
      Hint: Use this box to specify more
    </Hint>
    <ErrorMessage
      role="alert"
      visuallyHiddenText="Error: "
    >
      At least 20 characters...
    </ErrorMessage>
    <Input
      id="input-id4"
      type="text"
      width="10"
    />
  </ContextProvider>
</div>
`;

exports[`Form Group matches snapshot Form group label , Select and Error message 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
    <Hint>
      Hint: Use this box to specify more
    </Hint>
    <ErrorMessage
      role="alert"
      visuallyHiddenText="Error: "
    >
      At least 20 characters...
    </ErrorMessage>
    <Select>
      <Option>
        Select
      </Option>
      <Option
        value="1"
      >
        NHS.UK frontend option 1
      </Option>
      <Option
        value="2"
      >
        NHS.UK frontend option 2
      </Option>
      <Option
        value="3"
      >
        NHS.UK frontend option 3
      </Option>
    </Select>
  </ContextProvider>
</div>
`;

exports[`Form Group matches snapshot Form group label , Textarea and Error message 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
    <Hint>
      Hint: Use this box to specify more
    </Hint>
    <ErrorMessage
      role="alert"
      visuallyHiddenText="Error: "
    >
      At least 20 characters...
    </ErrorMessage>
    <Textarea
      id="no-ni-reason2"
      label="Why can't you provide a National Insurance number?"
      name="no-ni-reason2"
      rows={5}
    />
  </ContextProvider>
</div>
`;

exports[`Form Group matches snapshot Form group label and Textarea 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
    <Hint>
      Hint: Use this box to specify more
    </Hint>
    <Textarea
      id="no-ni-reason2"
      label="Why can't you provide a National Insurance number?"
      name="no-ni-reason2"
      rows={5}
    />
  </ContextProvider>
</div>
`;

exports[`Form Group matches snapshot Form group, label and Hint 1`] = `
<div
  className="nhsuk-form-group"
>
  <ContextProvider
    value={
      Object {
        "error": undefined,
        "inputID": undefined,
        "isInFormGroup": true,
        "setError": [Function],
        "setInputID": [Function],
      }
    }
  >
    <Label>
      Please specify (others)
    </Label>
    <Hint>
      Hint: Use this box to specify more
    </Hint>
  </ContextProvider>
</div>
`;
